File: go/parsers/challenge.go
package parsers

import (
	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/auth"
	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type Challenge struct {
	gen *gen.Challenge
}

func NewChallengeFields(g *gen.Challenge) messages.ChallengeFields {
	return &Challenge{gen: g}
}

func (c *Challenge) AuthMethod() string {
	val, _ := c.gen.AuthMethod()
	return val
}

func (c *Challenge) Extra() map[string]any {
	extra := make(map[string]any)
	challenge, err := c.gen.Challenge()
	if err == nil {
		extra["challenge"] = challenge
	}

	salt, err := c.gen.Salt()
	if err == nil {
		extra["salt"] = salt
		extra["iterations"] = c.gen.Iterations()
		extra["keylen"] = c.gen.Keylen()
	}
	return extra
}

func ChallengeToCapnproto(m *messages.Challenge) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	challenge, err := gen.NewRootChallenge(seg)
	if err != nil {
		return nil, err
	}

	challengeString, _ := m.Extra()["challenge"].(string)
	if err = challenge.SetChallenge(challengeString); err != nil {
		return nil, err
	}

	if err := challenge.SetAuthMethod(m.AuthMethod()); err != nil {
		return nil, err
	}

	if m.AuthMethod() == auth.MethodCRA {
		saltString, ok := m.Extra()["salt"].(string)
		if ok {
			if err = challenge.SetSalt(saltString); err != nil {
				return nil, err
			}
		}

		iterations, ok := m.Extra()["iterations"].(uint32)
		if ok {
			challenge.SetIterations(iterations)
		}

		keylen, ok := m.Extra()["keylen"].(uint16)
		if ok {
			challenge.SetKeylen(keylen)
		}
	}

	data, err := msg.MarshalPacked()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeChallenge, data, nil), nil
}

func CapnprotoToChallenge(data []byte) (*messages.Challenge, error) {
	msg, err := capnp.UnmarshalPacked(data)
	if err != nil {
		return nil, err
	}

	challenge, err := gen.ReadRootChallenge(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewChallengeWithFields(NewChallengeFields(&challenge)), nil
}

File: python/wamp_msgs_capnp/parsers/challenge.py
import os
from typing import Any
from pathlib import Path

import capnp
from wampproto import auth
from wampproto.messages.challenge import Challenge, IChallengeFields

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
module_file = os.path.join(root_dir, "schemas", "challenge.capnp")
challenge_capnp = capnp.load(str(module_file))


class ChallengeFields(IChallengeFields):
    def __init__(self, gen):
        self._gen = gen

    @property
    def authmethod(self) -> str:
        return self._gen.authMethod

    @property
    def extra(self) -> dict[str, Any]:
        extra = {}
        if self._gen.challenge:
            extra["challenge"] = self._gen.challenge

        if self._gen.salt:
            extra["salt"] = self._gen.salt
            extra["iterations"] = self._gen.iterations
            extra["keylen"] = self._gen.keylen

        return extra


def challenge_to_capnproto(c: Challenge) -> bytes:
    challenge = challenge_capnp.Challenge.new_message()

    challenge_string = c.extra.get("challenge", "")
    challenge.challenge = challenge_string
    challenge.authMethod = c.authmethod

    if c.authmethod == auth.WAMPCRAAuthenticator.TYPE:
        if c.extra.get("salt", None) is not None:
            challenge.salt = c.extra.get("salt")
        if c.extra.get("iterations", None) is not None:
            challenge.iterations = c.extra.get("iterations")
        if c.extra.get("keylen", None) is not None:
            challenge.keylen = c.extra.get("keylen")

    data = challenge.to_bytes_packed()

    return helpers.prepend_header(Challenge.TYPE, data, b"")


def capnproto_to_challenge(data: bytes) -> Challenge:
    message_data, _ = helpers.extract_message(data)
    challenge_obj = challenge_capnp.Challenge.from_bytes_packed(message_data)

    return Challenge(ChallengeFields(challenge_obj))
